### PR TITLE
[WFLY-12322] Avoid re-dispatching to the worker

### PR DESCRIPTION
**Pull request by fl4via**
Jira: https://issues.jboss.org/browse/WFLY-12322
EAP Jira (temporary): https://issues.jboss.org/browse/JBEAP-17049

We will have a proper JBEAP jira once we finish organizing the commits for JBEAP-17049

### CODE REVIEW

**Comment by soul2zimate** [comment]
The AsyncMethodTestCase failure looks related ? I also see it fails in downstream PR.

**Comment by fl4via** [comment]
@soul2zimate thanks for the feedback, I'll review this

**Comment by bmaxwell** [comment]
This change is causing a regression in QE's BeanPoolWithShortTimeoutTestCase.slsbWithPoolSize1AndShortTimeout - the test is EJB instance pool set to size 1 , with a timeout of 500 ms.  2 EJB requests sent at 200ms apart.  The 1st ejb request sleeps so that the 2nd request should timeout.  This change is not erroring out on the 2nd request, it allows the 2nd request to go ahead and complete.

**Comment by stuartwdouglas** [comment]
Retest this please

**Comment by stuartwdouglas** [comment]
This needs https://github.com/wildfly/jboss-ejb-client/pull/412 as this has had the unintended effect of making EJB requests on the same connection unable to run concurrently.

**Comment by soul2zimate** [comment]
eib-client has been upgraded to 4.0.23.Final which contains the fix in wildfly/jboss-ejb-client#412 @stuartwdouglas could you rebase this ?

**Comment by bstansberry** [comment]
retest this please

**Comment by bstansberry** [comment]
@fl4via @stuartwdouglas @soul2zimate @bmaxwell  where does this stand? I kicked off the security manager job again as it hadn't run correctly but that wasn't because if any problem I saw here. But reading the discussion I'm not clear on whether it's ready.

**Comment by soul2zimate** [comment]
Once those System.out.println() are removed, I believe this is good to merge.
I can't modify this @fl4via  @stuartwdouglas But I can open a new one for sure.

**Comment by soul2zimate** [comment]
Hi, @bmaxwell you mentioned regression in QE's BeanPoolWithShortTimeoutTestCase.slsbWithPoolSize1AndShortTimeout, How can we retest this ?

**Comment by bmaxwell** [comment]
@soul2zimate this was from a QE test suite https://gitlab.mw.lab.eng.bos.redhat.com/jbossqe-eap/tests-ejb/blob/master/ejb-client-qa-tests/src/test/java/org/jboss/qa/ejbclient/tests/beanpool/BeanPoolWithShortTimeoutTestCase.java#L63  when they were testing JBEAP-17268

**Comment by bstansberry** [comment]
retest this please

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/async/AsyncBean.java [hunk 1]
```diff
@@ -85,21 +86,28 @@ public void asyncMethod(CountDownLatch latch, CountDownLatch latch2) throws Inte
...
-        String result;
-        result = ctx.wasCancelCalled() ? "true" : "false";
-
-        synchronizeBean.latchCountDown();
-        long end = System.currentTimeMillis() + 5000;
-        while (System.currentTimeMillis() < end) {
-            if (ctx.wasCancelCalled()) {
-                break;
+        try {
+            System.out.println("ASYNC REMOTE ");
```
Comment (jamezp) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/async/AsyncBean.java]: This should probably be removed

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/async/AsyncBean.java [hunk 2]
```diff
@@ -85,21 +86,28 @@ public void asyncMethod(CountDownLatch latch, CountDownLatch latch2) throws Inte
...
+            String result;
+            result = ctx.wasCancelCalled() ? "true" : "false";
+
+            System.out.println("DO COUNTDOWN ");
```
Comment (jamezp) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/async/AsyncBean.java]: This should probably be removed

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/async/AsyncBeanSynchronizeSingleton.java [hunk 3]
```diff
@@ -40,6 +40,7 @@
     public void reset() {
         latch = new CountDownLatch(1);
         latch2 = new CountDownLatch(1);
+        System.out.println("RESET " + latch);
```
Comment (jamezp) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/async/AsyncBeanSynchronizeSingleton.java]: This should probably be removed

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/async/AsyncBeanSynchronizeSingleton.java [hunk 4]
```diff
@@ -51,6 +52,7 @@ public void latch2CountDown() {
     }
 
     public void latchAwaitSeconds(int sec) throws InterruptedException {
+        System.out.println("AWAIT " + latch);
```
Comment (jamezp) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/async/AsyncBeanSynchronizeSingleton.java]: This should probably be removed

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/async/AsyncBean.java [hunk 5]
```diff
@@ -85,21 +86,28 @@ public void asyncMethod(CountDownLatch latch, CountDownLatch latch2) throws Inte
...
-        }
-        result += ";";
+            result += ";";
 
-        result += ctx.wasCancelCalled() ? "true" : "false";
-        return new AsyncResult<String>(result);
+            result += ctx.wasCancelCalled() ? "true" : "false";
+            return new AsyncResult<String>(result);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (jamezp) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/async/AsyncBean.java]: This should probably be removed

**Comment by fl4via** [comment]
before cleaning the commit added by Stuart to investigate test failures, I'll try to run the QE test with this branch to see if the test is passing or not

**Comment by fl4via** [comment]
for the record... I updated the PR but forgot to add an observation: the test that @bmaxwell pointed passes with this branch.

**Pull request merged**